### PR TITLE
Support case sensitive column for `with_column_renamed`

### LINKED ISSUE #6845: dataframe.with_column_renamed() does not rename properly
### Describe the bug

The function Dataframe.with_column_renamed() cannot change uppercase single words to lower case words.
For instance,

```rust
df.with_column_renamed("Package", "package")?; // Produces "Package". Should be "package".
df.with_column_renamed("City Name", "city_name")?; // Produces "city_name". OK.
```

### To Reproduce

I have the following CSV table

|City Name|Package     |Variety    |Date   |Low Price|High Price|Mostly Low|
|---------|------------|-----------|-------|---------|----------|----------|
|BALTIMORE|24 inch bins|HOWDEN TYPE|9/24/16|160      |160       |160       |
|BALTIMORE|24 inch bins|HOWDEN TYPE|9/24/16|160      |160       |160       |
|BALTIMORE|24 inch bins|HOWDEN TYPE|11/5/16|90       |100       |90        |

I load the CVS data with this code:

```rust
let mut config = SessionConfig::new();
 config = config.with_information_schema(true);

let csv_options = CsvReadOptions::new()
  .has_header(true)
  .schema_infer_max_records(2000);

let mut df = context.read_csv("data/pumpkins.csv", csv_options).await?;
```

Then, I try to change the column names from uppercase to lower case and replacing white spaces with underscores using this code:

```rust
let field_names: Vec<String> = df.schema().fields().iter().map(|f| f.name().to_string()).collect::<Vec<String>>();
println!("{:?}", field_names);

let re = Regex::new(r"\s+").unwrap();

for field in field_names {
  let new_field = field.clone().trim().to_lowercase();
  let new_field = re.replace_all(&new_field, " ");
  let new_field: String = new_field.to_lowercase().trim().replace(" ", "_");

  println!("{} -> {}", field, new_field);

  df = df.with_column_renamed(field, &new_field)?;
}

df = df.with_column_renamed("Package", "package")?;

let field_names: Vec<String> = df.schema().fields().iter().map(|f| f.name().to_string()).collect::<Vec<String>>();

println!("{:?}", field_names);
```

And I get this results:

```text
["City Name", "Package", "Variety", "Date", "Low Price", "High Price", "Mostly Low"]
City Name -> city_name
Package -> package
Variety -> variety
Date -> date
Low Price -> low_price
High Price -> high_price
Mostly Low -> mostly_low
["city_name", "Package", "Variety", "Date", "low_price", "high_price", "mostly_low"]
```

As you can see, the function Dataframe.with_column_renamed() does not change the columns with an initial uppercase letter, even calling outside the for loop.

```rust
df = df.with_column_renamed("Package", "package")?;
```

### Expected behavior

I would like the function changes the column name regardless of the case

```rust
df = df.with_column_renamed("Package", "package")?;      // produces ‘package’
df = df.with_column_renamed("PACKAGE", "package")?; // produces ‘package’
```

I had to use the format function and escaping quotation marks in the for loop in order to get my desired output from the Dataframe.with_column_renamed() function.

```rust
// (...)
for field in field_names {
  //(...)
  df = df.with_column_renamed(format!("\"{}\"", field), &new_field)?;
}
// (...)
```

The dataframe.with_column_renamed() should change the column name properly. In my case, I want his output:

```text
["City Name", "Package", "Variety", "Date", "Low Price", "High Price", "Mostly Low"]
City Name -> city_name
Package -> package
Variety -> variety
Date -> date
Low Price -> low_price
High Price -> high_price
Mostly Low -> mostly_low
["city_name", "package", "variety", "date", "low_price", "high_price", "mostly_low"]
```

|city_name|package     |variety    |date   |low_price|high_price|mostly_low|
|---------|------------|-----------|-------|---------|----------|----------|
|BALTIMORE|24 inch bins|HOWDEN TYPE|9/24/16|160      |160       |160       |
|BALTIMORE|24 inch bins|HOWDEN TYPE|9/24/16|160      |160       |160       |
|BALTIMORE|24 inch bins|HOWDEN TYPE|11/5/16|90       |100       |90        |

I think it would be also useful to add a function to normalize all columns names to lower case and underscores

```rust
dataframe.to_lowercase() // produces ["city_name", "date", "low_price", "high_price"]
CsvReadOptions::new()::to_lowercase(true) // produces ["city_name", "date", "low_price", "high_price"]
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by comphead**
# Which issue does this PR close?

Closes #6845.

# Rationale for this change
Support case sensitive column renaming for dataframe API

# What changes are included in this PR?

# Are these changes tested?
Yes

# Are there any user-facing changes?

### CODE REVIEW

**Comment by comphead** [comment]
@Jefffrey appreciate if you have time to review

File: datafusion/core/src/dataframe.rs [hunk 1]
```diff
@@ -1014,12 +1014,21 @@ impl DataFrame {
     /// ```
     pub fn with_column_renamed(
         self,
-        old_name: impl Into<Column>,
+        old_name: &str,
```
Comment (Jefffrey) [inline datafusion/core/src/dataframe.rs:1017]: Could be

```suggestion
        old_name: impl Into<String>,
```

To maintain backwards compatibility technically?

**Review by Jefffrey — COMMENTED** [review]
Looks good, with a minor nitpick

Though I'm still unsure on using the `enable_ident_normalization` considering its technically a SQL Parser config, could be confusing for users as if using DataFrame API they might not expect such a config to affect them?

**Comment by comphead** [comment]
> Looks good, with a minor nitpick
> 
> Though I'm still unsure on using the `enable_ident_normalization` considering its technically a SQL Parser config, could be confusing for users as if using DataFrame API they might not expect such a config to affect them?

Changed the documentation to describe the change for users

**Comment by alamb** [comment]
CI is failing on master for reasons seemingly unrelated to this PR https://github.com/apache/arrow-datafusion/issues/7078

**Comment by alamb** [comment]
I merged this branch to main to pick up the fix for https://github.com/apache/arrow-datafusion/issues/7078. I plan to merge this PR in when CI passes

**Pull request merged**